alacritty/src/input.rs

```diff
@@ -735,6 +745,118 @@ impl<T: EventListener, A: ActionContext<T>> Processor<T, A> {
+            Touch::Zoom(zoom) => Touch::Invalid(zoom.slots()),
+            Touch::Scroll(event) | Touch::Select(event) => {
+                let mut set = HashSet::new();
+                set.insert(event.id);
+                Touch::Invalid(set)
+            },
+            Touch::Invalid(mut slots) => {
+                slots.insert(touch.id);
+                Touch::Invalid(slots)
+            },
+        };
+    }
+
+    /// Handle touch input movement.
+    pub fn on_touch_motion(&mut self, touch: TouchEvent) {
+        let touch_purpose = self.ctx.touch();
+        match touch_purpose {
+            Touch::None => (),
+            // Handle transition from tap to scroll/select.
+            Touch::Tap(start) => {
+                let delta_x = touch.location.x - start.location.x;
+                let delta_y = touch.location.y - start.location.y;
+                if delta_x.abs() > MAX_TAP_DISTANCE {
+                    // Update gesture state.
+                    let start_location = start.location;
+                    *touch_purpose = Touch::Select(*start);
+
+                    // Start simulated mouse input.
+                    self.mouse_moved(start_location);
+                    self.mouse_input(ElementState::Pressed, MouseButton::Left);
+
+                    // Apply motion since touch start.
+                    self.on_touch_motion(touch);
+                } else if delta_y.abs() > MAX_TAP_DISTANCE {
+                    // Update gesture state.
+                    *touch_purpose = Touch::Scroll(*start);
+
+                    // Apply motion since touch start.
+                    self.on_touch_motion(touch);
+                }
+            },
+            Touch::Zoom(zoom) => {
+                let font_delta = zoom.font_delta(touch);
+                self.ctx.change_font_size(font_delta);
+            },
+            Touch::Scroll(last_touch) => {
+                // Calculate delta and update last touch position.
+                let delta_y = touch.location.y - last_touch.location.y;
+                *touch_purpose = Touch::Scroll(touch);
+
+                self.scroll_terminal(0., delta_y * TOUCH_SCROLL_FACTOR);
+            },
+            Touch::Select(_) => self.mouse_moved(touch.location),
+            Touch::Invalid(_) => (),
+        }
+    }
+
+    /// Handle end of touch input.
+    pub fn on_touch_end(&mut self, touch: TouchEvent) {
+        // Finalize the touch motion up to the release point.
+        self.on_touch_motion(touch);
+
+        let touch_purpose = self.ctx.touch();
+        match touch_purpose {
+            // Simulate LMB clicks.
+            Touch::Tap(start) => {
+                let start_location = start.location;
+                *touch_purpose = Default::default();
+
+                self.mouse_moved(start_location);
+                self.mouse_input(ElementState::Pressed, MouseButton::Left);
+                self.mouse_input(ElementState::Released, MouseButton::Left);
```

Comment:
That should consider slots, otherwise you can 3-ple click with 3 fingers? So the tap should be counted from first/last insert or maybe when there's only a single finger at a time.

Reply:
As soon as a third finger is put on the touchscreen the invalid state is entered and you need to release all touch points to start any further touch input.